// react-intl/index.js
/*
* Copyright 2015, Yahoo Inc.
* Copyrights licensed under the New BSD License.
* See the accompanying LICENSE file for terms.
*/
import "@formatjs/ecma402-abstract";
import "@formatjs/intl";
import * as React from "react";
import { createFormattedComponent, createFormattedDateTimePartsComponent } from "./src/components/createFormattedComponent.js";
import { createIntl } from "./src/components/createIntl.js";
import FormattedDateTimeRange from "./src/components/dateTimeRange.js";
import injectIntl, { Context as IntlContext, Provider as RawIntlProvider } from "./src/components/injectIntl.js";
import FormattedMessage from "./src/components/message.js";
import FormattedPlural from "./src/components/plural.js";
import IntlProvider from "./src/components/provider.js";
import FormattedRelativeTime from "./src/components/relative.js";
import useIntl from "./src/components/useIntl.js";
import "./src/types.js";
export { createIntlCache, InvalidConfigError, MessageFormatError, MissingDataError, MissingTranslationError, IntlError as ReactIntlError, IntlErrorCode as ReactIntlErrorCode, UnsupportedFormatterError } from "@formatjs/intl";
export { createIntl, FormattedDateTimeRange, FormattedMessage, FormattedPlural, FormattedRelativeTime, injectIntl, IntlContext, IntlProvider, RawIntlProvider, useIntl };
export function defineMessages(msgs) {
	return msgs;
}
export function defineMessage(msg) {
	return msg;
}
// IMPORTANT: Explicit here to prevent api-extractor from outputing `import('./src/types').CustomFormatConfig`
export const FormattedDate = createFormattedComponent("formatDate");
export const FormattedTime = createFormattedComponent("formatTime");
export const FormattedNumber = createFormattedComponent("formatNumber");
export const FormattedList = createFormattedComponent("formatList");
export const FormattedDisplayName = createFormattedComponent("formatDisplayName");
export const FormattedDateParts = createFormattedDateTimePartsComponent("formatDate");
export const FormattedTimeParts = createFormattedDateTimePartsComponent("formatTime");
export { FormattedListParts, FormattedNumberParts } from "./src/components/createFormattedComponent.js";
